File: dotnet/Lexy.Poc/Lexy.Poc.Core/Language/ScenarioTable.cs
using System.Collections.Generic;
using Lexy.Poc.Core.Parser;

namespace Lexy.Poc.Core.Language
{
    public class ScenarioTable : ParsableNode
    {
        public TableHeaders Headers { get; private set; }
        public IList<TableRow> Rows { get; } = new List<TableRow>();

        public ScenarioTable(SourceReference reference) : base(reference)
        {
        }

        public override IParsableNode Parse(IParserContext context)
        {
            var line = context.CurrentLine;

            if (line.IsEmpty()) return this;

            if (Headers == null)
            {
                Headers = TableHeaders.Parse(context);
                return this;
            }

            var row = TableRow.Parse(context);
            if (row != null)
            {
                Rows.Add(row);
            }

            return this;
        }

        protected override IEnumerable<INode> GetChildren()
        {
            if (Headers != null)
            {
                yield return Headers;
            }

            foreach (var row in Rows)
            {
                yield return row;
            }
        }

        protected override void Validate(IValidationContext context)
        {
        }
    }
}

File: typescipt/src/language/scenarios/scenarioTable.ts


export class ScenarioTable extends ParsableNode {
   public TableHeader Header { get; private set; }
   public Array<TableRow> Rows = list<TableRow>(): new;

   public ScenarioTable(SourceReference reference) : base(reference) {
   }

   public override parse(context: IParseLineContext): IParsableNode {
     if (Header == null) {
       Header = TableHeader.Parse(context);
       return this;
     }

     let row = TableRow.Parse(context);
     if (row != null) Rows.Add(row);

     return this;
   }

   public override getChildren(): Array<INode> {
     if (Header != null) yield return Header;

     foreach (let row in Rows) yield return row;
   }

   protected override validate(context: IValidationContext): void {
   }
}
